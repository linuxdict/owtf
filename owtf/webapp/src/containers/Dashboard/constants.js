export const LOAD_ERRORS = "owtf/Dashboard/LOAD_ERRORS",
  LOAD_ERRORS_SUCCESS = "owtf/Dashboard/LOAD_ERRORS_SUCCESS",
  LOAD_ERRORS_ERROR = "owtf/Dashboard/LOAD_ERRORS_ERROR";

export const CREATE_ERROR = "owtf/Dashboard/CREATE_ERROR",
  CREATE_ERROR_SUCCESS = "owtf/Dashboard/CREATE_ERROR_SUCCESS",
  CREATE_ERROR_ERROR = "owtf/Dashboard/CREATE_ERROR_ERROR";

export const DELETE_ERROR = "owtf/Dashboard/DELETE_ERROR",
  DELETE_ERROR_SUCCESS = "owtf/Dashboard/DELETE_ERROR_SUCCESS",
  DELETE_ERROR_ERROR = "owtf/Dashboard/DELETE_ERROR_ERROR";

export const LOAD_SEVERITY = "owtf/Dashboard/LOAD_SEVERITY",
  LOAD_SEVERITY_SUCCESS = "owtf/Dashboard/LOAD_SEVERITY_SUCCESS",
  LOAD_SEVERITY_ERROR = "owtf/Dashboard/LOAD_SEVERITY_ERROR";

export const LOAD_TARGET_SEVERITY = "owtf/Dashboard/LOAD_TARGET_SEVERITY",
  LOAD_TARGET_SEVERITY_SUCCESS = "owtf/Dashboard/LOAD_TARGET_SEVERITY_SUCCESS",
  LOAD_TARGET_SEVERITY_ERROR = "owtf/Dashboard/LOAD_TARGET_SEVERITY_ERROR";