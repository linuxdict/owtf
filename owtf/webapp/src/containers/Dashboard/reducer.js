/*
 * ErrorReducer
 *
 * The reducer takes care of our data. Using actions, we can change our
 * application state.
 * To add a new action, add it to the switch statement in the reducer function
 *
 * Example:
 * case YOUR_ACTION_CONSTANT:
 *   return state.set('yourStateVariable', true);
 */

import { fromJS } from "immutable";
import { combineReducers } from "redux-immutable"; // combineReducers of 'redux' doesn't work with immutable.js

import {
  LOAD_ERRORS,
  LOAD_ERRORS_SUCCESS,
  LOAD_ERRORS_ERROR,
  CREATE_ERROR,
  CREATE_ERROR_SUCCESS,
  CREATE_ERROR_ERROR,
  DELETE_ERROR,
  DELETE_ERROR_SUCCESS,
  DELETE_ERROR_ERROR,
  LOAD_SEVERITY,
  LOAD_SEVERITY_SUCCESS,
  LOAD_SEVERITY_ERROR,
  LOAD_TARGET_SEVERITY,
  LOAD_TARGET_SEVERITY_SUCCESS,
  LOAD_TARGET_SEVERITY_ERROR,
} from "./constants";

// The initial state of the errors.
const initialErrorsState = fromJS({
  loading: true,
  error: false,
  errors: false
});

export function errorsLoadReducer(state = initialErrorsState, action) {
  switch (action.type) {
    case LOAD_ERRORS:
      return state
        .set("loading", true)
        .set("error", false)
        .set("errors", false);
    case LOAD_ERRORS_SUCCESS:
      return state
        .set("loading", false)
        .set("error", false)
        .set("errors", action.errors);
    case LOAD_ERRORS_ERROR:
      return state
        .set("loading", false)
        .set("error", action.error)
        .set("errors", false);
    default:
      return state;
  }
}

// The initial state of the error create
const initialCreateState = fromJS({
  loading: false,
  error: false
});

export function errorCreateReducer(state = initialCreateState, action) {
  switch (action.type) {
    case CREATE_ERROR:
      return state.set("loading", true).set("error", false);
    case CREATE_ERROR_SUCCESS:
      return state.set("loading", false).set("error", false);
    case CREATE_ERROR_ERROR:
      return state.set("error", action.error).set("loading", false);
    default:
      return state;
  }
}

// The initial state of the error delete
const initialDeleteState = fromJS({
  loading: false,
  error: false
});

export function errorDeleteReducer(state = initialDeleteState, action) {
  switch (action.type) {
    case DELETE_ERROR:
      return state.set("loading", true).set("error", false);
    case DELETE_ERROR_SUCCESS:
      return state.set("loading", false).set("error", false);
    case DELETE_ERROR_ERROR:
      return state.set("error", action.error).set("loading", false);
    default:
      return state;
  }
}

// The initial state of the severity.
const initialSeverityState = fromJS({
  loading: true,
  error: false,
  severity: false
});

export function severityLoadReducer(state = initialSeverityState, action) {
  switch (action.type) {
    case LOAD_SEVERITY:
      return state
        .set("loading", true)
        .set("error", false)
        .set("severity", false);
    case LOAD_SEVERITY_SUCCESS:
      return state
        .set("loading", false)
        .set("error", false)
        .set("severity", action.severity);
    case LOAD_SEVERITY_ERROR:
      return state
        .set("loading", false)
        .set("error", action.error)
        .set("severity", false);
    default:
      return state;
  }
}

// The initial state of the last target severity.
const initialTargetSeverityState = fromJS({
  loading: true,
  error: false,
  targetSeverity: false
});

export function targetSeverityLoadReducer(state = initialTargetSeverityState, action) {
  switch (action.type) {
    case LOAD_TARGET_SEVERITY:
      return state
        .set("loading", true)
        .set("error", false)
        .set("targetSeverity", false);
    case LOAD_TARGET_SEVERITY_SUCCESS:
      return state
        .set("loading", false)
        .set("error", false)
        .set("targetSeverity", action.targetSeverity);
    case LOAD_TARGET_SEVERITY_ERROR:
      return state
        .set("loading", false)
        .set("error", action.error)
        .set("targetSeverity", false);
    default:
      return state;
  }
}

export default combineReducers({
  load: errorsLoadReducer,
  create: errorCreateReducer,
  delete: errorDeleteReducer,
  loadSeverity: severityLoadReducer,
  loadTargetSeverity: targetSeverityLoadReducer,
});
