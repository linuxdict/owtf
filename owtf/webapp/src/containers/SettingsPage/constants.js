export const LOAD_CONFIGURATIONS = 'owtf/Configurations/LOAD_CONFIGURATIONS';
export const LOAD_CONFIGURATIONS_SUCCESS = 'owtf/Configurations/LOAD_CONFIGURATIONS_SUCCESS';
export const LOAD_CONFIGURATIONS_ERROR = 'owtf/Configurations/LOAD_CONFIGURATIONS_ERROR';

export const CHANGE_CONFIGURATIONS = 'owtf/Configurations/CHANGE_CONFIGURATIONS';
export const CHANGE_CONFIGURATIONS_SUCCESS = 'owtf/Configurations/CHANGE_CONFIGURATIONS_SUCCESS';
export const CHANGE_CONFIGURATIONS_ERROR = 'owtf/Configurations/CHANGE_CONFIGURATIONS_ERROR';
