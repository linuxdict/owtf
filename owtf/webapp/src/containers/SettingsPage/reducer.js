import { fromJS } from 'immutable';
import { combineReducers } from 'redux-immutable'; // combineReducers of 'redux' doesn't work with immutable.js

import {
  LOAD_CONFIGURATIONS,
  LOAD_CONFIGURATIONS_SUCCESS,
  LOAD_CONFIGURATIONS_ERROR,
  CHANGE_CONFIGURATIONS,
  CHANGE_CONFIGURATIONS_SUCCESS,
  CHANGE_CONFIGURATIONS_ERROR,
} from './constants';


// The initial state of the configurations.
const initialConfigurationState = fromJS({
  loading: true,
  error: false,
  configurations: false,
});

export function configurationsLoadReducer(state = initialConfigurationState, action) {
  switch (action.type) {
    case LOAD_CONFIGURATIONS:
      return state
        .set('loading', true)
        .set('error', false)
        .set('configurations', false);
    case LOAD_CONFIGURATIONS_SUCCESS:
      return state
        .set('error', false)
        .set('loading', false)
        .set('configurations', action.configurations);
    case LOAD_CONFIGURATIONS_ERROR:
      return state
        .set('loading', false)
        .set('error', action.error)
        .set('configurations', false);
    default:
      return state;
  }
}

// The initial state of the configurations change
const initialChangeState = fromJS({
  loading: false,
  error: false,
});

export function configurationsChangeReducer(state = initialChangeState, action) {
  switch (action.type) {
    case CHANGE_CONFIGURATIONS:
      return state
        .set('loading', true)
        .set('error', false);
    case CHANGE_CONFIGURATIONS_SUCCESS:
      return state
        .set('loading', false)
        .set('error', false);
    case CHANGE_CONFIGURATIONS_ERROR:
      return state
        .set('loading', false)
        .set('error', action.error);
    default:
      return state;
  }
}

export default combineReducers({
  load: configurationsLoadReducer,
  change: configurationsChangeReducer,
});
