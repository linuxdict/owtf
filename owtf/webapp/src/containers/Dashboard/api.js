import Request from "../../utils/request";
import { API_BASE_URL } from "../../utils/constants";

function getHeaders() {
  return {
    headers: {
      "Content-Type": "application/x-www-form-urlencoded;charset=UTF-8",
      Authorization: `Bearer ${localStorage.getItem("token")}`
    }
  };
}

export function getErrorsAPI() {
  const requestURL = `${API_BASE_URL}errors/`;
  // Call our request helper (see 'utils/request')
  const options = getHeaders();
  const request = new Request(requestURL, options);
  return request.get.bind(request);
}

export function postErrorAPI() {
  const requestURL = `${API_BASE_URL}errors/`;
  const options = getHeaders();
  const request = new Request(requestURL, options);
  return request.post.bind(request);
}

export function deleteErrorAPI(action) {
  const error_id = action.error_id.toString();
  const requestURL = `${API_BASE_URL}errors/${error_id}/`;
  const options = getHeaders();
  const req_options = {
    responseAs: "text",
    ...options
  };
  const request = new Request(requestURL, req_options);
  return request.delete.bind(request);
}

export function getSeverityAPI() {
  const requestURL = `${API_BASE_URL}dashboard/severitypanel/`;
  // Call our request helper (see 'utils/request')
  const options = getHeaders();
  const request = new Request(requestURL, options);
  return request.get.bind(request);
}

export function getTargetSeverityAPI() {
  const requestURL = `${API_BASE_URL}targets/severitychart/`;
  // Call our request helper (see 'utils/request')
  const options = getHeaders();
  const request = new Request(requestURL, options);
  return request.get.bind(request);
}
