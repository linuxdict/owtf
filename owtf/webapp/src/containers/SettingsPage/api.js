import Request from "../../utils/request";
import { API_BASE_URL } from "../../utils/constants";

function getHeaders() {
  return {
    headers: {
      "Content-Type": "application/x-www-form-urlencoded;charset=UTF-8",
      Authorization: `Bearer ${localStorage.getItem("token")}`
    }
  };
}

const getConfigs = () => {
  const requestURL = `${API_BASE_URL}configuration/`;
  // Call our request helper (see 'utils/request')
  const options = getHeaders();
  const request = new Request(requestURL, options);
  return request.get.bind(request);
};

const changeConfig = () => {
  const requestURL = `${API_BASE_URL}configuration/`;
  const options = getHeaders();
  const request = new Request(requestURL, options);
  return request.patch.bind(request);
};

export const fetchConfigAPI = getConfigs();
export const patchConfigAPI = changeConfig();
