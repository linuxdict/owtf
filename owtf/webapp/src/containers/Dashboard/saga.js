/**
 * Fetch, Create and Change the error from API
 */

import { call, put, takeLatest } from "redux-saga/effects";
import { LOAD_ERRORS, CREATE_ERROR, DELETE_ERROR, LOAD_SEVERITY, LOAD_TARGET_SEVERITY } from "./constants";
import {
  loadErrors,
  errorsLoaded,
  errorsLoadingError,
  errorCreated,
  errorCreatingError,
  errorDeleted,
  errorDeletingError,
  severityLoaded,
  severityLoadingError,
  targetSeverityLoaded,
  targetSeverityLoadingError,
} from "./actions";
import { getErrorsAPI, postErrorAPI, deleteErrorAPI, getSeverityAPI, getTargetSeverityAPI } from "./api";
import { Dashboard } from "./index";
import "@babel/polyfill";

const dashboard = new Dashboard();
/**
 * Fetch Errors request/response handler
 */
export function* getErrors() {
  const fetchAPI = getErrorsAPI();
  try {
    // Call our request helper (see 'utils/request')
    const errors = yield call(fetchAPI);
    yield put(errorsLoaded(errors));
  } catch (error) {
    yield put(errorsLoadingError(error));
  }
}

/**
 * Post Error request/response handler
 */
export function* postError(action) {
  const postAPI = postErrorAPI();
  try {
    yield call(postAPI, action.post_data);
    yield put(errorCreated());
    yield put(loadErrors());
  } catch (error) {
    yield put(errorCreatingError(error));
  }
}

/**
 * Delete Error request/response handler
 */
export function* deleteError(action) {
  const deleteAPI = deleteErrorAPI(action);
  try {
    yield call(deleteAPI);
    yield put(errorDeleted());
    yield put(loadErrors());
    dashboard.toasterSuccess(action.error_id, "deleteError");
  } catch (error) {
    yield put(errorDeletingError(error));
    dashboard.toasterError(error);
  }
}

/**
 * Fetch Severity request/response handler
 */
export function* getSeverity() {
  const fetchAPI = getSeverityAPI();
  try {
    // Call our request helper (see 'utils/request')
    const severity = yield call(fetchAPI);
    yield put(severityLoaded(severity.data));
  } catch (error) {
    yield put(severityLoadingError(error));
  }
}

/**
 * Fetch last target severity request/response handler
 */
export function* getTargetSeverity() {
  const fetchAPI = getTargetSeverityAPI();
  try {
    // Call our request helper (see 'utils/request')
    const targetSeverity = yield call(fetchAPI);
    yield put(targetSeverityLoaded(targetSeverity.data));
  } catch (error) {
    yield put(targetSeverityLoadingError(error));
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export default function* errorsSaga() {
  // Watches for LOAD_ERROR actions and calls getError when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  // It will be cancelled automatically on component unmount
  yield takeLatest(LOAD_ERRORS, getErrors);
  yield takeLatest(CREATE_ERROR, postError);
  yield takeLatest(DELETE_ERROR, deleteError);
  yield takeLatest(LOAD_SEVERITY, getSeverity);
  yield takeLatest(LOAD_TARGET_SEVERITY, getTargetSeverity);
}
