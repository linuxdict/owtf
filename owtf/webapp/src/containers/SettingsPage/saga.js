/**
 * Fetch, Create and Change the configurations from API
 */

import { call, put, takeLatest } from 'redux-saga/effects';
import { LOAD_CONFIGURATIONS, CHANGE_CONFIGURATIONS } from './constants';
import { configurationsLoaded, configurationsLoadingError, configurationsChangingError, configurationsChanged, loadConfigurations } from './actions';
import { fetchConfigAPI, patchConfigAPI } from './api';
import "@babel/polyfill";

/**
 * Fetch Configuration request/response handler
 */
export function* getConfigurations() {
  try {
    const configurations = yield call(fetchConfigAPI);
    yield put(configurationsLoaded(configurations.data));
  } catch (error) {
    yield put(configurationsLoadingError(error));
  }
}

/**
 * Fetch Configuration request/response handler
 */
export function* patchConfigurations(action) {
  const patch_data = action.configurations;
  try {
    yield call(patchConfigAPI, patch_data);
    yield put(configurationsChanged());
    yield put(loadConfigurations());
  } catch (error) {
    yield put(configurationsChangingError(error));
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export default function* configurationSaga() {
  // Watches for LOAD_CONFIGURATIONS actions and calls getConfigurations when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  // It will be cancelled automatically on component unmount
  yield takeLatest(LOAD_CONFIGURATIONS, getConfigurations);
  yield takeLatest(CHANGE_CONFIGURATIONS, patchConfigurations);
}
