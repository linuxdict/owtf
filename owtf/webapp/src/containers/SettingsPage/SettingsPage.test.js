import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import '../../setupTests';
import { findByTestAtrr, checkProps} from '../../utils/testUtils';
import configureStore from 'redux-mock-store';
import ConnectedSettingsPage, { SettingsPage } from './index';
import { fromJS } from 'immutable';

import { 
	configurationsLoaded,
	configurationsLoadingError,
	configurationsChanged,
	configurationsChangingError,
	loadConfigurations,
} from './actions';
import { CHANGE_CONFIGURATIONS, LOAD_CONFIGURATIONS, LOAD_CONFIGURATIONS_SUCCESS, LOAD_CONFIGURATIONS_ERROR, CHANGE_CONFIGURATIONS_SUCCESS, CHANGE_CONFIGURATIONS_ERROR } from './constants';

import { getConfigurations, patchConfigurations } from './saga';
import { runSaga } from 'redux-saga';
import * as api from './api';

import { configurationsLoadReducer, configurationsChangeReducer } from './reducer';

const setUp = (initialState={}) => {
	const mockStore = configureStore();
	const store = mockStore(initialState);
	const wrapper = shallow(<ConnectedSettingsPage store={store} />);
	return wrapper;
};

describe('Settings Page componemt', () => {

	describe('Testing dumb component', () => {

		let wrapper;
		let props;
		beforeEach(() => {
				props = {
					loading: false,
					fetchError: false,
					changeError: false,
					configurations: {},
					onFetchConfiguration: jest.fn(),
					onChangeConfiguration: jest.fn(),
				};
				wrapper = shallow(<SettingsPage {...props} />);
		});

		it('Should have correct prop-types', () => {

			const expectedProps = {
				loading: true,
				fetchError: false,
				changeError: false,
				configurations: {},
				onFetchConfiguration: () => {
	
				},
				onChangeConfiguration: () => {
	
				}
			};
			const propsErr = checkProps(SettingsPage, expectedProps)
			expect(propsErr).toBeUndefined();
	
		});
	
		it('Should render without errors', () => {
			const component = findByTestAtrr(wrapper, 'settingsPageComponent');
			expect(component.length).toBe(1);
			expect(toJson(component)).toMatchSnapshot();
		});

		it('Should contain two sub-components', () => {
			const tabsNav = wrapper.find('ConfigurationTabsNav');
			expect(tabsNav.length).toBe(1);
			const tabsContent = wrapper.find('ConfigurationTabsContent');
			expect(tabsContent.length).toBe(1);
		});

		it('Should call onChangeConfiguration on button click', () => {
			expect(props.onFetchConfiguration.mock.calls.length).toBe(1);
			expect(props.onChangeConfiguration.mock.calls.length).toBe(0);
			const button = findByTestAtrr(wrapper, 'changeBtn');
			button.simulate('click');
			expect(props.onChangeConfiguration.mock.calls.length).toBe(1);
		});
		
	});

	describe('Testing connected component', () => {
		let wrapper;
		let initialState;
		beforeEach(() => {
			const load = {
				loading: false,
				error: false,
				configurations: {
					"test section 1": [
						{
							descrip: false,
							dirty: false,
							key: "key 1",
							section: "test section 1",
							value: "test value 1",
						},
						{
							descrip: false,
							dirty: false,
							key: "key 2",
							section: "test section 1",
							value: "test value 2",
						}
					]
				}
			}
			const change = {
					loading: false,
					error: false
			}
			const configurations = {
					load: load,
					change: change
			}
			initialState = fromJS({
					configurations: configurations
			})
			wrapper = setUp(initialState);
		});
		
		it('Props should match the initial state', () => {
			const loadingProp = initialState.get('configurations').get('load').get('loading');
			const fetchErrorProp = initialState.get('configurations').get('load').get('error');
			const changeErrorProp = initialState.get('configurations').get('change').get('error');
			const configProp = initialState.get('configurations').get('load').get('configurations');

			expect(wrapper.props().loading).toEqual(loadingProp);
			expect(wrapper.props().fetchError).toEqual(fetchErrorProp);
			expect(wrapper.props().changeError).toEqual(changeErrorProp);
			expect(wrapper.props().configurations).toEqual(configProp);
		});
	});

	describe('Testing the sagas', () => {

		describe('Testing getConfiguration saga', ()=> {

			it('Should load and return configs in case of success', async () => {
				// we push all dispatched actions to make assertions easier
				// and our tests less brittle
				const dispatchedActions = [];
		
				// we don't want to perform an actual api call in our tests
				// so we will mock the fetchAPI api with jest
				// this will mutate the dependency which we may reset if other tests
				// are dependent on it
				const mockedConfigs = {
					status: "success",
					data: {
						"test section 1": [
							{
								descrip: false,
								dirty: false,
								key: "key 1",
								section: "test section 1",
								value: "test value 1",
							},
						]
					}
				};
				api.fetchConfigAPI = jest.fn(() => Promise.resolve(mockedConfigs));
		
				const fakeStore = {
						getState: () => ({ state: 'test' }),
						dispatch: action => dispatchedActions.push(action),
				};
		
				// wait for saga to complete
				await runSaga(fakeStore, getConfigurations).done;
		
				expect(api.fetchConfigAPI.mock.calls.length).toBe(1);
				expect(dispatchedActions).toContainEqual(configurationsLoaded(mockedConfigs.data));
			});
			
			it('Should handle config load errors in case of failure', async () => {
					const dispatchedActions = [];
			
					// we simulate an error by rejecting the promise
					// then we assert if our saga dispatched the action(s) correctly
					const error = 'API server is down';
					api.fetchConfigAPI = jest.fn(() => Promise.reject(error));
			
					const fakeStore = {
							getState: () => ({ state: 'test' }),
							dispatch: action => dispatchedActions.push(action),
					};
			
					await runSaga(fakeStore, getConfigurations).done;
					
					expect(api.fetchConfigAPI.mock.calls.length).toBe(1);
					expect(dispatchedActions).toContainEqual(configurationsLoadingError(error));
			});

		})

		describe('Testing patchConfigurations saga', ()=> {
			
			let fakeAction;
			beforeEach(() => {
				fakeAction = {
					type: CHANGE_CONFIGURATIONS,
    			configurations: {
						"test section 1": [
							{
								descrip: false,
								dirty: false,
								key: "key 1",
								section: "test section 1",
								value: "test value 1",
							},
						]
					}
				}
			});

			it('Should load and return configs in case of success', async () => {
				// we push all dispatched actions to make assertions easier
				// and our tests less brittle
				const dispatchedActions = [];
		
				// we don't want to perform an actual api call in our tests
				// so we will mock the fetchAPI api with jest
				// this will mutate the dependency which we may reset if other tests
				// are dependent on it

				api.patchConfigAPI = jest.fn(() => Promise.resolve());
		
				const fakeStore = {
						getState: () => ({ state: 'test' }),
						dispatch: action => dispatchedActions.push(action),
				};
		
				// wait for saga to complete
				await runSaga(fakeStore, patchConfigurations, fakeAction).done;
		
				expect(api.patchConfigAPI.mock.calls.length).toBe(1);
				expect(dispatchedActions).toContainEqual(configurationsChanged());
				expect(dispatchedActions).toContainEqual(loadConfigurations());
			});
			
			it('Should handle config load errors in case of failure', async () => {
					const dispatchedActions = [];
			
					// we simulate an error by rejecting the promise
					// then we assert if our saga dispatched the action(s) correctly
					const error = 'API server is down';
					api.patchConfigAPI = jest.fn(() => Promise.reject(error));
			
					const fakeStore = {
							getState: () => ({ state: 'test' }),
							dispatch: action => dispatchedActions.push(action),
					};
			
					await runSaga(fakeStore, patchConfigurations,fakeAction).done;
					
					expect(api.patchConfigAPI.mock.calls.length).toBe(1);
					expect(dispatchedActions).toContainEqual(configurationsChangingError(error));
			});

		})

	});

	describe('Testing reducers', () => {

		describe('Testing configurationsLoadReducer', ()=> {

			let initialState;
			beforeEach(() => {
				initialState = {
					loading: true,
					error: false,
					configurations: false,
				}
			});

			it('Should return the initial state', () => {
				const newState = configurationsLoadReducer(undefined, {});
				expect(newState.toJS()).toEqual(initialState);
			});

			it('Should handle LOAD_CONFIGURATIONS', () => {
				
				const newState = configurationsLoadReducer(undefined, {
					type: LOAD_CONFIGURATIONS,
				})
				expect(newState.toJS()).toEqual(initialState);

			});

			it('Should handle LOAD_CONFIGURATIONS_SUCCESS', () => {
				const configs = { testSection: [{
					descrip: false,
					dirty: false,
					key: "key 1",
					section: "test section 1",
					value: "test value 1",
				}]}

				const newState1 = configurationsLoadReducer(undefined, {
					type: LOAD_CONFIGURATIONS_SUCCESS,
					configurations: configs
				})
				expect(newState1.toJS()).toEqual({
					loading: false,
					error: false,
					configurations: configs
				})

				const newState2 = configurationsLoadReducer(fromJS({
					loading: true,
					error: true,
					configurations: true
				}), 
				{
					type: LOAD_CONFIGURATIONS_SUCCESS,
					configurations: configs
				})
				expect(newState2.toJS()).toEqual({
					loading: false,
					error: false,
					configurations: configs
				})
			})

			it('Should handle LOAD_CONFIGURATIONS_ERROR', () => {
				const newState1 = configurationsLoadReducer(undefined, {
					type: LOAD_CONFIGURATIONS_ERROR,
					error: 'Test error'
				})
				expect(newState1.toJS()).toEqual({
					loading: false,
					error: 'Test error',
					configurations: false
				})

				const newState2 = configurationsLoadReducer(fromJS({
					loading: true,
					error: true,
					configurations: true
				}), 
				{
					type: LOAD_CONFIGURATIONS_ERROR,
					error: 'Test error'
				})
				expect(newState2.toJS()).toEqual({
					loading: false,
					error: 'Test error',
					configurations: false
				})
			})	

		});

		describe('Testing configurationsChangeReducer', ()=> {

			let initialState;
			beforeEach(() => {
				initialState = {
					loading: false,
					error: false,
				}
			});

			it('Should return the initial state', () => {
				const newState = configurationsChangeReducer(undefined, {});
				expect(newState.toJS()).toEqual(initialState);
			});

			it('Should handle CHANGE_CONFIGURATIONS', () => {
				
				const newState = configurationsChangeReducer(undefined, {
					type: CHANGE_CONFIGURATIONS,
					configurations: {}
				})
				expect(newState.toJS()).toEqual({ loading: true, error: false});

			});

			it('Should handle CHANGE_CONFIGURATIONS_SUCCESS', () => {

				const newState1 = configurationsChangeReducer(undefined, {
					type: CHANGE_CONFIGURATIONS_SUCCESS,
				})
				expect(newState1.toJS()).toEqual({
					loading: false,
					error: false,
				})

				const newState2 = configurationsChangeReducer(fromJS({
					loading: true,
					error: true,
				}), 
				{
					type: CHANGE_CONFIGURATIONS_SUCCESS,
				})
				expect(newState2.toJS()).toEqual({
					loading: false,
					error: false,
				})
			})

			it('Should handle CHANGE_CONFIGURATIONS_ERROR', () => {
				const newState1 = configurationsChangeReducer(undefined, {
					type: CHANGE_CONFIGURATIONS_ERROR,
					error: 'Test error'
				})
				expect(newState1.toJS()).toEqual({
					loading: false,
					error: 'Test error',
				})

				const newState2 = configurationsChangeReducer(fromJS({
					loading: true,
					error: true,
				}), 
				{
					type: CHANGE_CONFIGURATIONS_ERROR,
					error: 'Test error'
				})
				expect(newState2.toJS()).toEqual({
					loading: false,
					error: 'Test error',
				})

			});	
		});
	});
});
