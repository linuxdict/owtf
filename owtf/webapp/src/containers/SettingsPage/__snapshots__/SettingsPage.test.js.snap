// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings Page componemt Testing dumb component Should render without errors 1`] = `
<withTheme(Pane)
  data-test="settingsPageComponent"
  display="flex"
  flexDirection="column"
  margin={20}
>
  <withTheme(Pane) />
  <withTheme(Pane)
    margin={20}
  >
    <withTheme(Button)
      appearance="primary"
      className="pull-right"
      data-test="changeBtn"
      disabled={true}
      onClick={[Function]}
    >
      Update Configuration!
    </withTheme(Button)>
  </withTheme(Pane)>
  <withTheme(Pane)
    display="flex"
    margin={20}
  >
    <ConfigurationTabsNav
      configurations={Object {}}
      handleTabSelect={[Function]}
      selectedIndex={-1}
    />
    <withTheme(Pane)
      flex="1"
      padding={16}
    >
      <ConfigurationTabsContent
        configurations={Object {}}
        handleConfigurationChange={[Function]}
        selectedIndex={-1}
      />
    </withTheme(Pane)>
  </withTheme(Pane)>
</withTheme(Pane)>
`;
